<h1> Soal 1: Pemahaman Tentang Model Evaluasi</h1>

Jawab pertanyaan di bawah ini dengan bahasa masing-masing?

1. Apa perbedaan antara data latih, data validasi, dan data test?
2. Bagaimana cara kita menilai performa suatu model?
3. Apa itu Confusion Matrix? Jelaskan secara lengkap!
4. Apa itu Classification Report dari sklearn?

Jawab:
1. Perbedaannya adalah data latih adalah data yang digunakan mesin untuk belajar, sedangkan data validasi adalah data yang digunakan untuk mengevaluasi model, dan data test adalah data yang digunakan sebagai prediksi.
2. Performa suatu model dapat di nilai dengan Memvalidasi model dan cross validation.
3. Confusion matrix adalah tabel yang digunakan untuk melihat performa dari suatu model atas suatu set data test.
4. Classification report dari sklearn adalah suatu module yang digunakan untuk mengukur kualitas dari prediksi model.

---

<h1>Soal 2: Aplikasi Model Evaluasi</h1>

Kali ini kita akan menggunakan data untuk memprediksi kelangsungan hidup pasien yang telah mengalami operasi payudara. Dengan informasi yang dimiliki terkait pasien, kita akan membuat model untuk memprediksi apakah pasien akan bertahan hidup dalam waktu lebih dari 5 tahun atau tidak.

Lebih Lengkapnya kalian bisa membaca informasi tentang dataset di link berikut: https://raw.githubusercontent.com/jbrownlee/Datasets/master/haberman.names

Buat model Klasifikasi (Model/Algoritma Bebas) untuk memprediksi status pasien dengan ketentuan sebagai berikut:

1. Bagi kedua data ini menjadi data training dan data test dengan test_size=0.25.
3. Pelajar tentang metrics roc_auc_score kemudian buatlah model dan evaluasi dengan menggunakan teknik cross-validation dengan scoring 'roc_auc'. Baca https://scikit-learn.org/stable/modules/generated/sklearn.model_selection.cross_val_score.html untuk menggunakan metric roc_auc saat cross-validation.
3. Berapa score rata2 dari model dengan teknik cross-validation tersebut?
4. Prediksi data test dengan model yang telah kalian buat!
5. Bagaimana hasil confusion matrix dari hasil prediksi tersebut?
6. Bagaimana classification report dari hasil prediksi tersebut?
5. Seberapa baik model anda dalam memprediksi seorang pasien mempunyai status positive?
6. Seberapa baik model anda dalam memprediksi seorang pasien mempunyai status negatif?

In [ ]:
import pandas as pd

url = 'https://raw.githubusercontent.com/jbrownlee/Datasets/master/haberman.csv'
list_cols = ['Age', "Patient's Years", "N_positive_ax", "survival_status"]
df = pd.read_csv(url, names=list_cols)

In [ ]:
df.head()

Age  Patient's Years  N_positive_ax  survival_status
0   30               64              1                1
1   30               62              3                1
2   30               65              0                1
3   31               59              2                1
4   31               65              4                1

In [ ]:
df['survival_status'].value_counts()

1    225
2     81
Name: survival_status, dtype: int64

In [ ]:
from sklearn.model_selection import train_test_split
from sklearn.neighbors import KNeighborsClassifier

x = df.drop('survival_status', axis=1)
y = df['survival_status']
knn = KNeighborsClassifier(n_neighbors=3, weights='distance')

In [ ]:
x_train, x_test, y_train, y_test = train_test_split(x,y, test_size=0.25, stratify=y, random_state=64)
knn.fit(x_train, y_train)
y_predict = knn.predict(x_test)

In [ ]:
from sklearn.model_selection import cross_val_score

cv_results = cross_val_score(knn, x, y, cv=3)
cv_results.mean()

0.6830065359477123

In [ ]:
from sklearn.metrics import roc_auc_score
roc_auc_score(y_test, y_predict)

0.6210526315789473

In [ ]:
from sklearn.metrics import confusion_matrix, classification_report
confusion_matrix(y_test, y_predict)

array([[48,  9],
       [12,  8]], dtype=int64)

In [ ]:
print(classification_report(y_test, y_predict))

              precision    recall  f1-score   support

           1       0.80      0.84      0.82        57
           2       0.47      0.40      0.43        20

    accuracy                           0.73        77
   macro avg       0.64      0.62      0.63        77
weighted avg       0.71      0.73      0.72        77



1. Seberapa baik model anda dalam memprediksi seorang pasien mempunyai status positive?
    - Model memprediksi dengan akurasi 82%
2. Seberapa baik model anda dalam memprediksi seorang pasien mempunyai status negatif?
    - Model memprediksi dengan akurasi 43%

---

<h1> Soal 3: Pemahaman Tentang Model Selection</h1>

Jelaskan dengan bahasa sendiri!

1. Apa itu Bias dan Variance?
2. Apa itu Overfitting dan Underfitting?
3. Apa yang bisa kita lakukan untuk mengatur kompleksitas dari model?
4. Bagaimana model yang baik?
5. Kapan kita menggunakan GridSearchcv dan kapan menggunakan RandomizedSearchCV?

Jawab:
1. Bias adalah perbedaan antara prediksi model dengan data yang sesungguhnya yang kita coba untuk prediksi, error bias terjadi jika data kita terlalu simpel. Variance adalah variabilitas dari model prediksi yang memberi tahu kita persebaran data, error variance terjadi jika data kita terlalu kompleks.
2. Overfitting adalah kondisi dimana model terlalu kompleks dan memprediksi data training dengan sangat baik tetapi tidak bisa memprediksi data test dengan baik. Sedangkan underfitting adalah kondisi dimana model terlalu simpel sehingga sistem kurang pintar.
3. Agar bisa mengatur kompleksitas dari model, cara yang dilakukan adalah mengatur hyperparameter dan menggunakan cross validation.
4. Model yang baik adalah model yang tidak terlalu simpel ataupun kompleks sehingga tidak terjadi over maupun underfitting, Menggunakan hyperparameter optimal dan memiliki akurasi tinggi.
5. Kita menggunakan GridSearchCV dan RandomizedSearchCV untuk mencari hyperparameter dengan akurasi terbaik sehingga bisa membuat model yang optimum. GridSearchCV mencari nilai terbaik dari semua kombinasi yang mungkin dengan konsekuensi yaitu membutuhkan waktu yang lama tetapi RandomizedSearchCV mencari nilai terbaik dari kombinasi secara acak sehingga membutuhkan waktu yang relatif singkat.

---

<h1> Soal 4: Aplikasi Model Selection</h1>

1. Bagi kedua data berikut ini menjadi data training dan data test dengan test_size=0.25.
2. Gunakan algoritma KNN sebagai model classifier.
3. Gunakan fungsi GridSearchCV untuk hyperparameter tuning dan model selection.
4. jumlah fold bebas!, gunakan scoring 'roc_auc'
5. Definisikan kombinasi hyperparameter untuk model selection dengan GridSearchCV. kombinasi Hyperparameter bebas, baca lagi dokumentasi KNN di link berikut https://scikit-learn.org/stable/modules/generated/sklearn.neighbors.KNeighborsClassifier.html untuk memahami lagi jenis2 hyperparameter di algorithma KNN.
6. Latih model terhadap data training.
7. Apa hyperparameter terbaik untuk kombinasi hyperparameter kalian?
8. Berapa score validasi terbaik dari model tersebut?
9. Prediksi probabilitasi output dari model yang telah di buat terhadap data test. note : gunakan method .predict_proba() untuk menghasilkan output probabilitas
10. Perhatikan bahwa hasil prediksi ada 2, dimana masing2 adalah nilai probabilitas untuk setiap class label. Ambil nilai probabilitas pasien phositive meninggal dalam waktu kurang dari 5 tahun. note : gunakan bantuan attirubte .classes_ untuk mengetahui urutan label dari hasil prediksi probabilitas.
11. Berapa nilai score roc_auc untuk data test?
12. Apakah model anda termasuk baik, overtting, atau underfitting?
13. Ulangi tahap di atas namun kali ini menggunakan algoritma DecisionTreeClassifier dan kalian bisa menggunakan RandomizedSearchCV apabila process training lama. pelajari algoritma DecisionTreeClassifier di linkberikut: https://scikit-learn.org/stable/modules/generated/sklearn.tree.DecisionTreeClassifier.html?highlight=decisiontreeclassifier#sklearn.tree.DecisionTreeClassifier
14. Bandingkan scorenya dengan Algoritma KNN, mana yang lebih baik?

Note : Data Science adalah experiment, sangat di dimungkinkan memerlukan beberapa kali percobaan untuk mendapatkan hasil yang terbaik! Happy Coding :)

In [ ]:
import pandas as pd

url = 'https://raw.githubusercontent.com/jbrownlee/Datasets/master/haberman.csv'
list_cols = ['Age', "Patient's Years", "N_positive_ax", "survival_status"]
df = pd.read_csv(url, names=list_cols)

In [ ]:
df.head()

Age  Patient's Years  N_positive_ax  survival_status
0   30               64              1                1
1   30               62              3                1
2   30               65              0                1
3   31               59              2                1
4   31               65              4                1

In [ ]:
from sklearn.neighbors import KNeighborsClassifier
from sklearn.model_selection import GridSearchCV
from sklearn.model_selection import train_test_split
import numpy as np

X = df.drop('survival_status', axis=1)
y = df['survival_status']

X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.25, random_state=21)

model = KNeighborsClassifier()

In [ ]:
param_grid = {'n_neighbors':np.arange(3,50), 'weights':['distance', 'uniform'],
              'algorithm':['auto', 'ball_tree', 'kd_tree', 'brute'], 'leaf_size':[20,30,40]}
gscv = GridSearchCV(model, param_grid=param_grid, scoring='accuracy', cv=5)
gscv.fit(X,y)

GridSearchCV(cv=5, estimator=KNeighborsClassifier(),
             param_grid={'algorithm': ['auto', 'ball_tree', 'kd_tree', 'brute'],
                         'leaf_size': [20, 30, 40],
                         'n_neighbors': array([ 3,  4,  5,  6,  7,  8,  9, 10, 11, 12, 13, 14, 15, 16, 17, 18, 19,
       20, 21, 22, 23, 24, 25, 26, 27, 28, 29, 30, 31, 32, 33, 34, 35, 36,
       37, 38, 39, 40, 41, 42, 43, 44, 45, 46, 47, 48, 49]),
                         'weights': ['distance', 'uniform']},
             scoring='accuracy')

In [ ]:
gscv.best_params_

{'algorithm': 'ball_tree',
 'leaf_size': 20,
 'n_neighbors': 31,
 'weights': 'uniform'}

In [ ]:
gscv.best_score_

0.7745637228979375

In [ ]:
knn = KNeighborsClassifier(n_neighbors=37, weights='uniform')
_x = knn.fit(X_train, y_train)
_x

KNeighborsClassifier(n_neighbors=37)

In [ ]:
y_pred = knn.predict(X_test)

In [ ]:
from sklearn.metrics import roc_auc_score
roc_auc_score(y_test, y_pred)

0.5919471153846154

In [ ]:
_x.predict_proba(X_test)

array([[0.81081081, 0.18918919],
       [0.81081081, 0.18918919],
       [0.75675676, 0.24324324],
       [0.86486486, 0.13513514],
       [0.72972973, 0.27027027],
       [0.48648649, 0.51351351],
       [0.81081081, 0.18918919],
       [0.48648649, 0.51351351],
       [0.75675676, 0.24324324],
       [0.72972973, 0.27027027],
       [0.83783784, 0.16216216],
       [0.72972973, 0.27027027],
       [0.75675676, 0.24324324],
       [0.54054054, 0.45945946],
       [0.72972973, 0.27027027],
       [0.78378378, 0.21621622],
       [0.7027027 , 0.2972973 ],
       [0.89189189, 0.10810811],
       [0.7027027 , 0.2972973 ],
       [0.72972973, 0.27027027],
       [0.78378378, 0.21621622],
       [0.64864865, 0.35135135],
       [0.78378378, 0.21621622],
       [0.78378378, 0.21621622],
       [0.86486486, 0.13513514],
       [0.81081081, 0.18918919],
       [0.81081081, 0.18918919],
       [0.91891892, 0.08108108],
       [0.83783784, 0.16216216],
       [0.78378378, 0.21621622],
       [0.

In [ ]:
_x.classes_

array([1, 2], dtype=int64)

In [ ]:
from sklearn.model_selection import cross_validate
score = cross_validate(knn, X, y, cv=5, return_train_score=True)

In [ ]:
score['test_score'].mean()

0.7713379164463248

In [ ]:
score['train_score'].mean()

0.7663466042154566

Model saya baik karena train score dengan accuracy 76.63%% dan test score dengan accuracy 77.13%

---

## DecisionTreeClassifier

In [ ]:
from sklearn.tree import DecisionTreeClassifier
from sklearn.model_selection import RandomizedSearchCV

In [ ]:
dtc = DecisionTreeClassifier()

param_rs = {'random_state':np.arange(0,60), 'criterion':['gini', 'entropy'], 'splitter':['best', 'random']}
rscv= RandomizedSearchCV(dtc,param_rs, n_iter=60 , scoring='accuracy', cv=5)
rscv.fit(X,y)

RandomizedSearchCV(cv=5, estimator=DecisionTreeClassifier(), n_iter=60,
                   param_distributions={'criterion': ['gini', 'entropy'],
                                        'random_state': array([ 0,  1,  2,  3,  4,  5,  6,  7,  8,  9, 10, 11, 12, 13, 14, 15, 16,
       17, 18, 19, 20, 21, 22, 23, 24, 25, 26, 27, 28, 29, 30, 31, 32, 33,
       34, 35, 36, 37, 38, 39, 40, 41, 42, 43, 44, 45, 46, 47, 48, 49, 50,
       51, 52, 53, 54, 55, 56, 57, 58, 59]),
                                        'splitter': ['best', 'random']},
                   scoring='accuracy')

In [ ]:
rscv.best_score_

0.7059756742464305

In [ ]:
rscv.best_params_

{'splitter': 'random', 'random_state': 36, 'criterion': 'gini'}

In [ ]:
print('Score KNearestNeighbors: ', gscv.best_score_)
print('Score DecisionTreeClassifier: ',rscv.best_score_)

Score KNearestNeighbors:  0.7745637228979375
Score DecisionTreeClassifier:  0.7059756742464305
